### ISSUE TITLE
TransformerEncoderLayer always warns when using src_key_padding_mask in inference mode.

### ISSUE THREAD

**Issue by jhseu** [pytorch/pytorch#100469]
### 🐛 Describe the bug

```python

import torch

with torch.inference_mode():
    d_model = 4
    layer = torch.nn.TransformerEncoderLayer(d_model, 2, 2, batch_first=True)
    layer.eval()
    x = torch.randn(5, 10, d_model)
    pad = torch.rand(5, 10) > 0.5
    layer(x, src_key_padding_mask=pad)
```

Gives the warning:
```
/home/jhseu/.local/lib/python3.10/site-packages/torch/nn/modules/transformer.py:668: UserWarning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly. (Triggered internally at ../aten/src/ATen/native/transformers/attention.cpp:152.)
  return torch._transformer_encoder_layer_fwd(
```

This is because the layer is converting it to a float mask regardless of what the user passes, and then ATen warns about requiring a bool mask. This warning appears on every call when doing inference with libtorch.

### Versions

Collecting environment information...
PyTorch version: 2.0.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.25.0
Libc version: glibc-2.37

Python version: 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201] (64-bit runtime)
Python platform: Linux-6.2.13-arch1-1-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 530.41.03
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.8.0
/usr/lib/libcudnn_adv_infer.so.8.8.0
/usr/lib/libcudnn_adv_train.so.8.8.0
/usr/lib/libcudnn_cnn_infer.so.8.8.0
/usr/lib/libcudnn_cnn_train.so.8.8.0
/usr/lib/libcudnn_ops_infer.so.8.8.0
/usr/lib/libcudnn_ops_train.so.8.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          32
On-line CPU(s) list:             0-31
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen Threadripper 1950X 16-Core Processor
CPU family:                      23
Model:                           1
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        1
Frequency boost:                 enabled
CPU(s) scaling MHz:              68%
CPU max MHz:                     3400.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        6789.07
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev
Virtualization:                  AMD-V
L1d cache:                       512 KiB (16 instances)
L1i cache:                       1 MiB (16 instances)
L2 cache:                        8 MiB (16 instances)
L3 cache:                        32 MiB (4 instances)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.1
[pip3] pytorch-lightning==2.0.1
[pip3] torch==2.0.0+cu118
[pip3] torchaudio==2.0.1+cu118
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.15.1+cu118
[pip3] triton==2.0.0
[conda] Could not collect

cc @jbschlosser @albanD @mruberry @walterddr @mikaylagawarecki @bhosmer @cpuhrsch @erichan1 @drisspg

**Comment by janeyx99** [comment]
It does look like https://github.com/pytorch/pytorch/blob/843ead134ca6f168df98912efaeea72a0b380909/torch/nn/functional.py#L5042 this line always converts a mask to float. @mikaylagawarecki are you familiar with why this is the case?

Regardless, I do agree this is a confusing warning message that should be fixed.

**Comment by mikekgfb** [comment]
Almost all scenarios use a float mask adding -infinity to mask out elements.
It might be most expedient to just remove the warning.  (We might think how we change the mask or the mask usage for inference, depending what the best representation is. This might be expedient to include at the time when we refactor the current C++ implementation.)

**Comment by mshavliuk** [comment]
The warning is printed only once and can easily be filtered out by `warnings.filterwarnings("ignore")` unless the model is compiled. Otherwise they get printed for every single layer of the transformer, on every inference run, which is extremely distractive and slows down the process.

Any clues how to disable them for a cpp backend?

```
[W attention.cpp:150] Warning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly. (function masked_softmax)
[W attention.cpp:150] Warning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly. (function masked_softmax)
[W attention.cpp:150] Warning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly. (function masked_softmax)
```

**Comment by mikekgfb** [comment]
At a minimum we should just remove the warning from the CPP execution path.  

> The warning is printed only once and can easily be filtered out by warnings.filterwarnings("ignore") unless the model is compiled. Otherwise they get printed for every single layer of the transformer, on every inference run, which is extremely distractive and slows down the process.

What layer and overall model structure are you using?  src_key_padding_mask should be folded into a Nested Tensor in TransformerEncoder, so that downstream layers can execute with variable length inputs .  

This is happening here in transformer.py => https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/transformer.py#L315

**Comment by mikekgfb** [comment]
> Hm the behavior in that line seems aligned with the documentation for [key_padding_mask here](https://github.com/pytorch/pytorch/blob/843ead134ca6f168df98912efaeea72a0b380909/torch/nn/functional.py#L5096-L5098)
> 
> This warning seems to come from [masked_softmax](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/attention.cpp#L150-L152). Perhaps a bool mask is needed in some paths (e.g. need_weights) and not others
> 
> cc @drisspg

Exactly this!  I think step 1 is removing the symptom, not the cause.  In principle we could check whether we're going into native_* cpp code (same conditions and make producing canonical masks produce Bool canonical masks if we're on the trajectory to call native* cpp code.  Since we've discussing some refactoring of those paths in the future, that might be the right time to fix this holistically, and right now we can just fix the nuisance of a warning that the user can't plausibly fix.
